contracts/amm/contract/src/lib.rs

```diff
@@ -143,125 +143,66 @@ impl AmmContract {
-                    );
-                }
-                amount
-            }
-            _ => {
-                // Check the blobs are TransferFrom
-                return Err("Transfer blobs do not call the correct function".to_string());
-            }
-        };
+                sender: self.caller().0.clone(),
+                recipient: self.contract_name.0.clone(),
+                amount: from_amount,
+            },
+        )?;
 
         // Compute x,y and check swap is legit (x*y=k)
         let normalized_pair = UnorderedTokenPair::new(pair.0.clone(), pair.1.clone());
         let is_normalized_order = pair.0 <= pair.1;
-        match self.state.pairs.get_mut(&normalized_pair) {
-            Some((prev_x, prev_y)) => {
-                if is_normalized_order {
-                    let amount_b = *prev_y - (*prev_x * *prev_y / (*prev_x + from_amount));
-                    if amount_b != to_amount {
-                        return Err(format!(
-                            "Swap formula is not respected: {} * {} != {} * {}",
-                            (*prev_x + from_amount),
-                            (*prev_y - to_amount),
-                            prev_x,
-                            prev_y,
-                        ));
-                    }
-                    *prev_x += from_amount;
-                    *prev_y -= to_amount;
-                } else {
-                    let amount_b = *prev_x - (*prev_y * *prev_x / (*prev_y + from_amount));
-                    if amount_b != to_amount {
-                        return Err(format!(
-                            "Swap formula is not respected: {} * {} != {} * {}",
-                            (*prev_y + from_amount),
-                            (*prev_x - to_amount),
-                            prev_y,
-                            prev_x
-                        ));
-                    }
-                    *prev_y += from_amount;
-                    *prev_x -= to_amount;
-                }
-            }
-            None => {
-                return Err(format!("Pair {:?} not found in AMM state", pair));
-            }
-        };
+        let Some((prev_x, prev_y)) = self.state.pairs.get_mut(&normalized_pair) else {
+            return Err(format!("Pair {:?} not found in AMM state", pair));
+        };
+        let expected_to_amount = if is_normalized_order {
+            let k = *prev_y - (*prev_x * *prev_y / (*prev_x + from_amount));
+            *prev_x += from_amount;
+            *prev_y -= k; // we need to remove the full amount to avoid slipping
+            k
+        } else {
+            let k = *prev_x - (*prev_y * *prev_x / (*prev_y + from_amount));
+            *prev_y += from_amount;
+            *prev_x -= k; // we need to remove the full amount to avoid slipping
+            k
+        };
+
+        // Assert that we transferred less than that, within 2%
+        if to_amount < expected_to_amount * 100 / 102 {
```

Comment:
J'aurais fait comme ça 
```suggestion
        if to_amount < expected_to_amount * 98 / 100 {
```
je dis n'imp ? :thinking:

Reply:
Ça doit jouer sur les arrondis mais pas sûr que ça change tant de choses en pratique

Reply:
On est pas sur un slippage à 2% mais 1.96% si je me trompe pas
Mais osef aha

Reply:
je pense qu'un fix possible serait d'encadrer to_amount 
```suggestion
        let epsilon = 2/100;
        if  expected_to_amount * (1-epsilon) < to_amount && to_amount < expected_to_amount (1+epsilon) {
```

Par contre je suis pas certain qu'il n'y ait pas un scénario où la valeur de k diminue lentement si l'arrondi se fait "dans le mauvais sens"